oggm/workflow.py

```diff
@@ -34,20 +34,33 @@ def _init_pool():
     return mp.Pool(cfg.PARAMS['mp_processes'], initializer=_init_pool_globals, initargs=(cfg_contents,))
 
 
+def _merge_dicts(*dicts):
+    r = {}
+    for d in dicts:
+        r.update(d)
+    return r
+
+
 class _pickle_copier(object):
     """Pickleable alternative to functools.partial,
     Which is not pickleable in python2 and thus doesn't work
     with Multiprocessing."""
 
-    def __init__(self, func, **kwargs):
+    def __init__(self, func, gdir_has_kwargs, **kwargs):
         self.call_func = func
+        self.gdir_has_kwargs = gdir_has_kwargs
         self.out_kwargs = kwargs
 
     def __call__(self, gdir):
-        return self.call_func(gdir, **self.out_kwargs)
+        if self.gdir_has_kwargs:
+            gdir, gdir_kwargs = gdir
+            gdir_kwargs = _merge_dicts(self.out_kwargs, gdir_kwargs)
+            return self.call_func(gdir, **gdir_kwargs)
+        else:
+            return self.call_func(gdir, **self.out_kwargs)
 
 
-def execute_entity_task(task, gdirs, **kwargs):
+def execute_entity_task(task, gdirs, gdir_has_kwargs=False, **kwargs):
     """Execute a task on gdirs.
```

Comment:
I think it would be cleaner to remove the optional ``**kwargs`` as a whole, and have following function signature:

```python

def execute_entity_task(task, gdirs, kwargs=None):
    """Execute a task on gdirs.

    If you asked for multiprocessing, it will do it.

    Parameters
    ----------
    task: function
        the entity task to apply
    gdirs: list
        the list of oggm.GlacierDirectory to process
    kwargs: None, dict, or list of dict
        the keyword arguments to pass to the task. Set to a dict if the kwargs
        are the same for all instances of the task, and to a list of dict
        to pass different arguments to each instance (in that case, len(kwargs) 
        must be equal to len(gdirs))
    """
```

What do you think? We should be careful to see if this doesn't brake some code elsewhere

```